framework/src/materials/Material.C

```diff
@@ -171,8 +175,26 @@ Material::getOutputs()
 void
 Material::computeProperties()
 {
-  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+  // If this Material has the _constant_on_elem flag set, we take the
```

Comment:
Can we provide a virtual function `preComputeProperties()` to give users a chance to set up things needed by `computeQpProperties()`, for example doing some average over all qps for the expensive `constant_on_elem` calculations?